File: ZYH.WebControl_V/TmkControl/Js/ThemeSelector.js
function Class_ThemeSelector(ClientID, CallbackOptions, Settings) {
    var Self = this;
    Self.HidenField = $('#' + ClientID + '_hd');
    Self.Control = $('#' + ClientID);
    Self.Settings = Settings;
    Self.Themes;
    Self.Xml;
    Self.SelectedTheme;

    Self.Serialize = function () {
        var xml = '<ThemeSelector>';
        for (var i = 0; i < Self.Themes.length; i++) {
            xml += '<TH>';
            var Selected = Self.Themes[i].Selected; if (Selected) xml += '<S>true</S>';
            var Text = Self.Themes[i].Text; if (Text != '') xml += '<T>' + htmlEncode(Text) + '</T>';
            var Value = Self.Themes[i].Value; if (Value != '') xml += '<V>' + htmlEncode(Value) + '</V>';
            if (!Self.Themes[i].Enabled) xml += '<E>' + Self.Themes[i].Enabled + '</E>';
            var CssClass = Self.Themes[i].CssClass; if (CssClass != '') xml += '<C>' + CssClass + '</C>';
            var CssClass_Selected = Self.Themes[i].CssClass_Selected; if (CssClass_Selected != '') xml += '<CS>' + CssClass_Selected + '</CS>';
            xml += '</TH>';
        }
        xml += '</ThemeSelector>';
        return xml;
    }

    Self.Deserialize = function () {
        var DOM = XMLtoDOM(Self.Xml);
        Self.Themes = new Array();
        var Themes = $('TH', DOM);
        for (var i = 0; i < Themes.length; i++) {
            var t = new Object();
            t.ClientId = $('ID', Themes[i]).text();
            t.Text = $('T', Themes[i]).text();
            t.Value = $('V', Themes[i]).text();
            var E = $('E', Themes[i]).text(); if (E == '') { t.Enabled = true; } else { t.Enabled = (E.toLowerCase() == 'true'); }
            t.Selected = $('S', Themes[i]).text().toLowerCase() == 'true';
            t.CssClass = $('C', Themes[i]).text();
            t.CssClass_Selected = $('CS', Themes[i]).text();
            Self.Themes[i] = t;
            if (t.Selected) { Self.SelectedTheme = t; }
        }
    }

    Self.Init = function () {
        Self.Xml = $.base64Decode(Self.HidenField.val());
        Self.Deserialize();
    }

    Self.ThemeSelected = function (Value) {
        UpdateSelection(Value);
        //Call Server
        var FunBeforeCall = null;
        var FunAfterCall = null;
        var e_BeforeCall = { 'CancelCall': !Self.Settings.ServerEvents_ThemeSelected, 'ArgFromClient': '', 'CommandArg': Self.Settings.CommandArg, 'Instance': Self, 'EventTriggerHierarchyID': Self.Settings.EventTriggerHierarchyID, 'CallbackOptions': CallbackOptions };
        e_BeforeCall.ControlReserved = Self.Xml;
        e_BeforeCall.SelectedThemeValue = Value;
        if (Self.Settings.ClientEvent_BeforeCallback_ThemeSelected != '') { FunBeforeCall = eval(Self.Settings.ClientEvent_BeforeCallback_ThemeSelected); }
        if (Self.Settings.ClientEvent_AfterCallback_ThemeSelected != '') { FunAfterCall = eval(Self.Settings.ClientEvent_AfterCallback_ThemeSelected); }
        eval(Self.Settings.CallbackManagerClientInstance).CallServer(Self.Settings.UniqueID, ClientID, '20', FunBeforeCall, e_BeforeCall, Self.AfterCallProcessing, FunAfterCall);
    }

    Self.AfterCallProcessing = function (id, e_AfterCall) {
        switch (e_AfterCall.CallbackOptions.AfterCallAction) {
            case 0: //'NA':
            case 11: //'RecallToSynchro':
                var xml = Self.Serialize();
                if (e_AfterCall.Context.e_BeforeCall.ControlReserved != xml) {
                    e_AfterCall.Context.e_BeforeCall.ControlReserved = xml;
                    e_AfterCall.ReFiredCallback = true;
                    e_AfterCall.CallbackManager_ClientInstance.DelayReFireCallback(e_AfterCall.Context);
                    return;
                }
                else {
                    RenderSelection();
                }
                break;
            case 1: //'RefreshToSynchro':
                Self.Xml = e_AfterCall.ControlReserved;
                Self.HidenField.val($.base64Encode(Self.Xml));
                Self.Deserialize();
                RenderSelection();
                break;
        }
        for (var i = 0; i < Self.Themes.length; i++) {
            if (Self.Themes[i].Selected) { e_AfterCall.SelectedThemeValue = Self.Themes[i].Value; return; }
        }
    }

    function UpdateSelection(Value) {
        for (var i = 0; i < Self.Themes.length; i++) {
            Self.Themes[i].Selected = false;
            if (Self.Themes[i].Value == Value) { Self.Themes[i].Selected = true; }
        }
        Self.Xml = Self.Serialize();
        Self.HidenField.val($.base64Encode(Self.Xml));
    }

    function RenderSelection() {
        for (var i = 0; i < Self.Themes.length; i++) {
            var Theme = $('#' + Self.Themes[i].ClientId, Self.Control); Theme.attr('class', Self.Themes[i].CssClass);
            if (Self.Themes[i].Selected) { Theme.attr('class', Self.Themes[i].CssClass_Selected); }
        }
    }
}

File: ZYH.WebControl_V/TmkControl/ThemeSelector.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.ComponentModel;
using System.Drawing;
using System.Web.UI.WebControls;
using System.Web.UI;

namespace ZYH.WebControl_V.TmkControl
{
    [DefaultEvent("ThemeSelected"), ToolboxBitmap(typeof(ThemeSelector), "Image.ToolBox.ThemeSelector.gif")]
    public class ThemeSelector : CallbackControl_Base
    {
        protected HiddenField _HiddenField;
        ThemeSelectorSetting _Settings = new ThemeSelectorSetting();
        ThemeSelectorData _Data = new ThemeSelectorData();

        public ThemeSelector() : base(System.Web.UI.HtmlTextWriterTag.Div) { }

        #region Event
        public delegate void ThemeSelectedHandler(object sender, ThemeSelectorEventArgs e);
        public event ThemeSelectedHandler ThemeSelected;

        protected virtual void OnThemeSelected(ThemeSelectorEventArgs e)
        {
            if (ThemeSelected != null) ThemeSelected(this, e);
        }

        public override void OnRaiseCallbackEvent(CallbackParameters CallbackParameters, CallbackResult CallbackResult)
        {
            ThemeSelectorEventArgs e = new ThemeSelectorEventArgs();
            e.CallbackManager = this.CallbackManager;
            e.ArgFromClient = CallbackParameters.ArgFromClient;
            e.Command = CallbackParameters.Command;
            e.CommandArg = this.CommandArg;
            e.CallbackOptions = this._CallbackOptions;
            var SelectedTheme = this.Themes.Where(x => x.Selected).FirstOrDefault();
            if (SelectedTheme != null) e.SelectedThemeValue = SelectedTheme.Value;
            switch (e.Command)
            {
                case CallbackCommands.ThemeSelector_ThemeSelected:
                    OnThemeSelected(e);
                    break;
            }
            if (SelectedTheme != null && SelectedTheme.Value != e.SelectedThemeValue)
            {
                this.Themes.ForEach(x => x.Selected = false);
                this.Themes.Where(x => x.Value == e.SelectedThemeValue).First().Selected = true;
            }
            CallbackResult.ArgToClient = e.ArgToClient;
            CallbackResult.CallbackOptions = e.CallbackOptions;
            CallbackResult.ControlReserved = _Data.ToXML(this.ClientID);
        }
        #endregion

        #region Property
        [Browsable(true), Category("Behavior"), PersistenceMode(PersistenceMode.InnerProperty), RefreshProperties(RefreshProperties.All)]
        public ThemeCollection Themes
        {
            get { return _Data.Themes; }
        }

        [Browsable(false)]
        public Theme SelectedTheme
        {
            get
            {
                return Themes.SelectedTheme;
            }
        }

        [Category("Behavior"), DefaultValue(""), Description("function BeforeCall(SenderID, e_BeforeCall) {} var e_BeforeCall = new Object(); e_BeforeCall.CancelCall. = false; e_BeforeCall.ArgFromClient = ''; e_BeforeCall.ControlReserved = ControlReserved; e_BeforeCall.Control = control")]
        public string ClientEvent_BeforeCallback_SelectedIndexChanged
        {
            get { return _Settings.ClientEvent_BeforeCallback_ThemeSelected; }
            set { _Settings.ClientEvent_BeforeCallback_ThemeSelected = value; }
        }

        [Category("Behavior"), DefaultValue(""), Description("function AfterCall(SenderID, e_AfterCall) {} //var e_AfterCall = new Object(); e_AfterCall.ControlReserved = ControlReserved; e_AfterCall.ArgToClient = ArgToClient;")]
        public string ClientEvent_AfterCallback_SelectedIndexChanged
        {
            get { return _Settings.ClientEvent_AfterCallback_ThemeSelected; }
            set { _Settings.ClientEvent_AfterCallback_ThemeSelected = value; }
        }

        [Category("Behavior"), DefaultValue("not applicable"), Description("not applicable.")]
        public new string ClientEvent_BeforeCallback { get { return "not applicable"; } }

        [Category("Behavior"), DefaultValue("not applicable"), Description("not applicable.")]
        public new string ClientEvent_AfterCallback { get { return "not applicable"; } }
        #endregion

        private void InitPostData()
        {
            string st = Page.Request.Params[_HiddenField.UniqueID];
            if (!string.IsNullOrEmpty(st))
            {
                var xml = System.Text.Encoding.UTF8.GetString(Convert.FromBase64String(st));
                _Data = SerializableObjectBase.CreateFromXML<ThemeSelectorData>(xml);
            }
        }

        protected override void OnInit(EventArgs e)
        {
            _HiddenField = new HiddenField(); _HiddenField.ID = "hd"; this.Controls.Add(_HiddenField);
            base.OnInit(e);
            if (Page.IsPostBack)
            {
                string xml;
                if (Page.IsCallback)
                {
                    if (CallbackManager.CallbackID == null) CallbackManager.InitCallbackParams();
                    if (CallbackManager.CallbackID == UniqueID)
                    {
                        xml = CallbackManager.CallbackParameters.ControlReserved;
                        _Data = SerializableObjectBase.CreateFromXML<ThemeSelectorData>(xml);
                    }
                    else if (CallbackManager.CallbackParameters.CallbackOptions.PostDataContainer != "NA")
                    {
                        InitPostData();
                    }
                }
                else
                {
                    InitPostData();
                }

            }
        }

        protected override void OnBeforeRender()
        {
            if (!DesignMode)
            {
                _Settings.CallbackManagerClientInstance = CallbackManager.ClientInstanceName;
                _Settings.UniqueID = this.UniqueID;
                _Settings.EventTriggerHierarchyID = this.EventTriggerHierarchyID;
                _Settings.CommandArg = this.CommandArg;

                _Settings.ServerEvents_ThemeSelected = (ThemeSelected == null) ? false : true;

                _HiddenField.Value = Convert.ToBase64String(new System.Text.UTF8Encoding().GetBytes(_Data.ToXML(this.ClientID)));

                string st = "";
                if (!CallbackManager.Scripts.IsScriptRegistered("ThemeSelectorJs")) { CallbackManager.Scripts.RegisterClientScriptInclude("ThemeSelectorJs", Page.ClientScript.GetWebResourceUrl(this.GetType(), "ZYH.WebControl_V.TmkControl.Js.ThemeSelector.js")); }
                st += "var " + ClientID + "_Instance = new Class_ThemeSelector('" + ClientID + "'," + base._CallbackOptions.ToJson_Js() + "," + _Settings.ToJson_Js() + ");";
                st += ClientID + "_Instance.Init();";
                CallbackManager.Scripts.RegisterStartupScript(ClientID + "InitJs", st);
            }
        }

        protected override void RenderContents(HtmlTextWriter writer)
        {
            if (this.Themes.Count == 0)
            {
                if (DesignMode) writer.Write("[" + this.ID + "]");
            }
            else
            {
                foreach (Theme t in this.Themes)
                {
                    writer.AddAttribute(HtmlTextWriterAttribute.Id, t.ClientId);
                    writer.AddAttribute(HtmlTextWriterAttribute.Class, t.Selected ? t.CssClass_Selected : t.CssClass);
                    writer.AddAttribute(HtmlTextWriterAttribute.Href, "javascript:" + ClientID + "_Instance.ThemeSelected('" + t.Value + "');");
                    writer.RenderBeginTag(HtmlTextWriterTag.A);
                    writer.RenderBeginTag(HtmlTextWriterTag.Span);
                    writer.Write(t.Text);
                    writer.RenderEndTag();
                    writer.RenderEndTag();
                    writer.WriteLine();
                }
            }
        }

        public override void RenderEndTag(HtmlTextWriter writer)
        {
            base.RenderEndTag(writer);
            if (!DesignMode)
            {
                writer.WriteLine();
                writer.Write(RenderControl(_HiddenField));
            }
        }
    }
}
